### PR TITLE
PR 4: (#150) Implementing Unit Tests for Lessons Model (Part 2)

### PR DESCRIPTION
**Pull request by EthanZ23**
## **Implementing Unit Testing for Lessons Model (Quiz, Question, Choice, Solution)**  

---

### **What does this PR do?**  
_This PR adds unit tests for the Lessons app API endpoints, specifically targeting the Quiz, Question, Choice, and Solution models. It also introduces the missing views, serializers, and URLs for the Quiz and Solution components to ensure complete coverage. These tests validate the correctness of CRUD operations and enhance the stability of the Quiz, Question, Choice, and Solution models._  

---

### **What has been added or changed?**  
✔ **Unit Test Implementation**:
- Added several tests to validate CRUD operations for Quiz, Question, Choice, and Solution models.
- Verified and ensured that responses are consistent with expected behavior.

✔ **Quiz Views, Serializers, and URLs**: 
- Implemented missing Quiz views, serializers, and URLs to better test coverage for lesson-related models.

✔ **Solution Views, Serializers, and URLs**:
- Integrated missing Solution views, serializers, and URLs to better test coverage for lesson-related models.

✔ **Test Framework Setup**:
- Extended the test infrastructure and added reusable test logic to support modular testing.

---

### **What problem does this solve?**  
_These additions improve our test coverage for the lessons-related API endpoints and help ensure consistent, predictable behavior as the lessons application evolves. It also closes out the remaining scope outlined in issue #150 and complements previous testing done for Assignments._  

---

## **Related Issue**  
🔗 **This PR addresses** [Issue #150 - _Continue implementing unit tests for the lessons app API endpoints, focusing on Quizzes, Questions, Choices, and Solutions._]

---

### **How was this feature developed?**  

- Created & added necessary Quiz views, serializers, and URLs.
- Added test cases for Quiz GET, POST, PATCH, and DELETE operations.
- Added test cases for Question GET, POST, PATCH, and DELETE operations.
- Added test cases for Choice GET, POST, PATCH, and DELETE operations.
- Created & added Solution views, serializers, and URLs.
- Added test cases for Solution GET, POST, PATCH, and DELETE operations.
- Validated correct status codes, responses, and data integrity.
- Used Django’s APIClient and pytest for request simulation, validation and test execution.

---

## **Design Documentation (If Applicable)**  
**Diagrams & Visuals** _(If the PR involves a major feature, link UML diagrams)_  
  
- **Use Case Diagram**: Can be added later if requested.

---

### **Step-by-step testing guide**  

1️⃣Navigate to `backend` directory. 
2️⃣Enable virtual environment. For Mac/Linux (bash/zsh) users, this is `source env/bin/activate`. For Windows (PowerShell) users, this is `.\env\Scripts\Activate.ps1`
3️⃣(If needed) Install dependencies: `pip install -r requirements.txt`
4️⃣Apply migrations: `python manage.py migrate`
5️⃣Run tests: `python manage.py test apps.lessons`
6️⃣Confirm all 21 tests pass successfully. 

---

## **Files Added/Modified**  
📂 **List of important files changed in this PR**  

| **File Name**      | **Description** |
|--------------------|----------------|
| `backend/apps/lessons/tests.py`    |  Implemented various test Case implementations |
| `backend/apps/lessons/urls.py`    | Registered new routes for Quiz and Solution |
| `backend/apps/lessons/views.py`        | Added CRUD views for Quiz and Solution models |
| `backend/apps/lessons/serializers.py`        | Defined serializers for Quiz and Solution models |

---

## **AI Code Usage**  
🤖 **Was AI-generated code used in this PR?**  

- [x] Yes (_Used to help structure logic in Choice PUT test case (reloading updated choices from the DB and validating updates)._)  
- [ ] No   

---

## **Checklist for Reviewers**  
**Things to check before approving this PR:**  
- [x] Does the feature work as expected?   
- [x] Is the code **clean, modular, and well-documented**?  
- [x] Do all tests pass successfully?
- [x] Are tests comprehensive and cover all (or most) CRUD operations? 
  
---

### **What improvements can be made after merging?**  
- Expand test coverage to handle more edge cases and error scenarios (validation errors).
- Include performance and integration tests for higher coverage. 

---

### **Final Notes**  
**Thanks for reviewing!** Please leave feedback if you have any suggestions or concerns.

-- Ethan

### CODE REVIEW

**Review by ChrisCRL — APPROVED** [review]
### 🔍 Peer Review: Assignment System API Tests
Hey Ethan👋 — Here's my feedback:
---
## ✅ What Looks Great
- Test coverage is extensive, covering all CRUD operations for Assignments, Quizzes, Questions, Choices, and Solutions.
- Authentication handling is properly implemented in the setUp method, ensuring tests run with proper permissions.
- Test organization is excellent, with clear delineation between different model test sections.
- Verification steps go beyond simply checking response status codes - you're verifying data integrity and confirming deletion.
---
## 💡 Suggestions for Improvement
- Consider adding negative test cases (e.g., attempting operations without proper authentication or permissions).
- Add tests for validation errors (e.g., creating questions with invalid data, choices without required fields).
---
## 🧪 Tested This Feature
I ran the following test steps:
- ✅ Reviewed all test cases for proper setup and teardown
- ✅ Verified test coverage across all models in the assignment system
- ✅ Confirmed proper test isolation to prevent test interdependence
- ✅ Checked that all API endpoints are properly tested
- ✅ Validated that permissions are correctly tested where applicable
---
### 🚀 Final Thoughts
The tests are well-structured and maintain a good balance between thoroughness and readability. With the addition of some negative test cases and validation testing, this model will be perfect. Approved :) SEND ITTTTTTTT (after the 2nd reviewer)

File: backend/apps/lessons/tests.py [hunk 1]
```diff
@@ -23,30 +23,46 @@ def setUp(self):
...
+        # Define URLs - used for assignment testing
         self.detail_url = reverse('assignment-detail', kwargs={'pk': self.assignment.pk})
         self.create_url = reverse('assignment-create')
 
-        # hardcode?
+        # Unless you want to hardcode:
         # self.list_url = '/lessons/assignments/'
-        #self.detail_url = f'/lessons/assignments/{self.assignment.pk}/'
-        #self.create_url = '/lessons/assignments/create/'
+        # self.detail_url = f'/lessons/assignments/{self.assignment.pk}/'
```
[the remark this answers is not in the document]
Comment (EthanZ23) [reply backend/apps/lessons/tests.py:56]: Thanks for pointing this out! I had originally left it as a heads-up for reviewers, but I’ll go ahead and remove it moving forward.

File: backend/apps/lessons/models.py [hunk 2]
```diff
@@ -151,19 +152,22 @@ def bulk_update_choices(cls, choices_data, question):
...
-        if self.question.question_type == 'MC':
-            correct_choices = ", ".join(choice.choice_text for choice in self.choices.all())
-            return f"Solution for Question {self.question.id}: {correct_choices}"
-        elif self.question.question_type == 'SA':
-            return f"Solution for Question {self.question.id}: {self.short_answer_text[:30]}"
-        else:
-            return f"Solution for Question {self.question.id}: (No Answer Set)"
+        quiz_title = self.question.quiz.assignment.title if self.question.quiz and self.question.quiz.assignment else "Unknown Quiz"
+        # (debugging) try-except can be removed once confident there's no circular logic
+        try:
```
[the remark this answers is not in the document]
Comment (EthanZ23) [reply backend/apps/lessons/models.py:165]: After I finish connecting with the frontend and running a few user tests, I’ll gladly remove it if it turns out to be unnecessary and not needed.

File: backend/apps/lessons/models.py [hunk 3]
```diff
@@ -112,7 +112,8 @@ class Choice(models.Model):  # Choices: Store the possible answer choices for mu
     is_correct = models.BooleanField(default=False)
 
     def __str__(self):
-        return f"Choice: {self.choice_text} ({'Correct' if self.is_correct else 'Incorrect'})"
+        quiz_title = self.question.quiz.assignment.title if self.question.quiz and self.question.quiz.assignment else "Unknown Quiz"
```
Comment (cj-ballesteros) [inline backend/apps/lessons/models.py:115]: nice edge case handling here
